File: 41-first-missing-positive/first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        // check if 1 already present & set out of bounds num to 1
        bool onePresent = false;
        int n = nums.size();

        for(int i = 0; i < n; i++){
            if (nums[i] == 1){
                onePresent = true;
            }
            else if (nums[i] > n || nums[i] < 1){
                nums[i] = 1;
            }
        }

        if (!onePresent) return 1;

        // mark presence of num by setting nums[num - 1] = negative
        for(int i = 0; i < n; i++){
            if (nums[abs(nums[i]) - 1] < 0){
                continue;
            }
            nums[abs(nums[i]) - 1] *= -1;
        }

        // if nums[idx] > 0 => idx + 1 was absent
        for(int i = 0; i < n; i++){
            if(nums[i] > 0){
                return i + 1;
            }
        }

        return n + 1;

    }
};

File: 41-first-missing-positive/first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        """
        3 4 -1 1

        cycle sort 
        3 4 -1 1
        1 4 3 -1
        """
        i = 0
        n = len(nums)

        while i < len(nums):
            correct_idx = nums[i] - 1
            if 1 <= nums[i] <= n and nums[i] != nums[correct_idx]:
                nums[i], nums[correct_idx] = nums[correct_idx], nums[i]
            else:
                i += 1

        for i in range(n):
            if nums[i] != i + 1:
                return i + 1
        
        return n + 1